pkg/trace/writer/trace.go

```diff
@@ -148,65 +156,49 @@ func (w *TraceWriter) Stop() {
 	w.sender.Stop()
 }
 
-func (w *TraceWriter) handleSampledTrace(sampledTrace *TracePackage) {
-	if sampledTrace == nil || sampledTrace.Empty() {
+func (w *TraceWriter) handleSampledTrace(pkg *TracePackage) {
+	if pkg == nil || pkg.Empty() {
 		log.Debug("Ignoring empty sampled trace")
 		return
 	}
 
-	trace := sampledTrace.Trace
-	events := sampledTrace.Events
+	apiTrace, size := traceutil.APITrace(pkg.Trace)
+	size += pb.Trace(pkg.Events).Msgsize()
 
-	n := len(trace) + len(events)
-
-	if w.spansInBuffer > 0 && w.spansInBuffer+n > w.conf.MaxSpansPerPayload {
-		// If we have data pending and adding the new data would overflow max spans per payload, force a flush
-		w.flushDueToMaxSpansPerPayload()
+	if w.payloadSize > 0 && w.payloadSize+size > payloadFlushThreshold {
+		// this new package would push us over, flush
+		log.Debug("Flushing... Reached size threshold.")
+		w.flush()
 	}
 
-	w.appendTrace(sampledTrace.Trace)
-	w.appendEvents(sampledTrace.Events)
-
-	if n > w.conf.MaxSpansPerPayload {
-		// If what we just added already goes over the limit, report this but lets carry on and flush
-		atomic.AddInt64(&w.stats.SingleMaxSpans, 1)
-		w.flushDueToMaxSpansPerPayload()
+	log.Tracef("Handling new trace with %d spans: %v", len(pkg.Trace), pkg.Trace)
+	w.traces = append(w.traces, apiTrace)
+	if len(pkg.Events) > 0 {
+		log.Tracef("Handling new APM events: %v", pkg.Events)
+		w.events = append(w.events, pkg.Events...)
 	}
-}
-
-func (w *TraceWriter) appendTrace(trace pb.Trace) {
-	numSpans := len(trace)
+	w.payloadSize += size
+	w.spansInBuffer += len(pkg.Trace) + len(pkg.Events)
 
-	if numSpans == 0 {
-		return
+	if size > payloadMaxSize {
+		log.Warnf("Trace might be dropped by API; size exceeds maximum (%d): %d", payloadMaxSize, size)
 	}
-
-	log.Tracef("Handling new trace with %d spans: %v", numSpans, trace)
-
-	w.traces = append(w.traces, traceutil.APITrace(trace))
-	w.spansInBuffer += numSpans
-}
-
-func (w *TraceWriter) appendEvents(events []*pb.Span) {
-	for _, event := range events {
-		log.Tracef("Handling new APM event: %v", event)
-		w.events = append(w.events, event)
+	if size > payloadFlushThreshold {
```

Comment:
Nit: may be clearer to move this check to happen next to `if w.payloadSize > 0 && w.payloadSize+size > payloadFlushThreshold {` above instead of after the flush call.

Reply:
Not sure what you mean by this, but I think that would result in a different outcome.

Reply:
Yeah it only makes sense if we get rid of the `flush()` then the code is only for counting how many times this happens. If we keep the `flush()` then doing it here is better.

Reply:
Let's see about suggestions on this https://github.com/DataDog/datadog-agent/pull/3364#discussion_r278169404 and based on that we'll make changes or not.